### DISCUSSION TITLE
Add `parent` argument to error condition

### DISCUSSION THREAD

**Discussion by dipterix** [ropensci/targets#967] [Ideas]
Currently whenever targets run into errors, the error messages aren't often informative although it prints out the last messages:

```
Error running targets::tar_make()
  Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
  Tips: https://books.ropensci.org/targets/debugging.html
  Last error:  ...
```

I was wondering if the run-time error condition can be passed as `parent` of the `tar_error`? For example like the followings

```r
tar_throw_run <- function (..., parent = NULL) {
  tar_error(message = paste0(...), class = c("tar_condition_run", 
                                             "tar_condition_targets"),
            parent = parent)
}

tar_error <- function (message, class, parent = NULL) {
  message <- cli::col_red(message)
  withr::local_options(list(rlang_backtrace_on_error = "none"))
  rlang::abort(message = message, class = class, call = tar_empty_envir, parent = parent)
}

callr_error <- function(condition, fun) {
  message <- sprintf(
    paste0(
      "Error running targets::%s()\n",
      "  Target errors: ",
      "targets::tar_meta(fields = error, complete_only = TRUE)\n",
      "  Tips: https://books.ropensci.org/targets/debugging.html\n",
      "  Last error: %s"
    ),
    fun,
    conditionMessage(condition)
  )
  tar_throw_run(message, parent = condition)
}
```

In this way, `callr_error` will be able to pass the "actual" error condition, including the errors all the way back.

**Comment by wlandau** [comment]
It's a good thought, but I'm not sure we need it for `targets`. `parent` does give us information about the original instigating error message:

```r
# Without parent
try_fetch(
  stop("original message"),
  error = function(condition) rlang::abort("new message")
)
#> Error in `handlers[[1L]]()`:
#> ! new message
#> Run `rlang::last_error()` to see where the error occurred.

# With parent
try_fetch(
  stop("original message"),
  error = function(condition) rlang::abort("new message", parent = condition)
)
#> Error:
#> ! new message
#> Caused by error in `withCallingHandlers()`:
#> ! original message
#> Run `rlang::last_error()` to see where the error occurred.
```

But `targets` already forwards the original error message even though it also adds stuff on top.

```r
# _targets.R
targets::tar_target(x, stop("original message"))
``` 

```r
# R console
tar_make()
#> • start target x
#> ✖ error target x
#> • end pipeline [0.09 seconds]
#> Error:
#> ! Error running targets::tar_make()
#>   Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
#>  Tips: https://books.ropensci.org/targets/debugging.html
#>   Last error: original message
```

When I tried forwarding the parent to `tar_throw_error()`, I `"original message"` printed twice.

**Reply by dipterix** [reply]
I guess my point is error messages are not always informative. In such cases, traceback might be more important in order to ping-point the bugs. Lots of common errors are pretty hard to debug in the context of targets. This is OK when debugging locally, but when deployed in production, we can't always obtain the outputs/meta from our customers due to confidential issues, hence it's a headache when customers send us error messages like the followings without further back-traces. 

```
Error in if (NULL) { : argument is of length zero
Error: attempt to apply non-function
``` 

In R >= 4.2, users can set rlang as global error handlers, hence the traceback will be more informative. Also I would avoid using `try_fetch` as it's experimental and has lots of limits (for example, only works in R>=4.2, and it's not using tryCatch, instead, it's using `withCallingHandlers`)

> When I tried forwarding the parent to tar_throw_error(), I saw "original message" printed twice.

I'm not sure what exactly happened here. Normally that shouldn't happen. My guess is you might have accidentally added `"original message"` error condition twice as a parent somewhere. (purely guess...)

**Reply by wlandau** [reply]
You can get a traceback if you save a [workspace](https://books.ropensci.org/targets/debugging.html#workspaces).

```r
# _targets.R
# _targets.R
library(targets)
tar_option_set(
  packages = c("broom", "broom.mixed", "dplyr", "nlme", "tibble", "tidyr"),
  workspace_on_error = TRUE # Every errored target saves a workspace file.
)

simulate_data <- function(units) {
  tibble(unit = seq_len(units), factor = rnorm(n = units, mean = 3)) %>%
    expand_grid(measurement = seq_len(4)) %>%
    mutate(outcome = sqrt(factor) + rnorm(n()))
}

analyze_data <- function(data) {
  gls(
    model = outcome ~ factor,
    data = data,
    correlation = corSymm(form = ~ measurement | unit),
    weights = varIdent(form = ~ 1 | measurement)
  ) %>%
    tidy(conf.int = TRUE, conf.level = 0.95)
}

list(
  tar_target(name = dataset1, command = simulate_data(100)),
  tar_target(name = model, command = analyze_data(dataset1))
)
```

```r
# R console
library(targets)
tar_make()
#> ✔ skip target dataset1
#> • start target model
#> • record workspace model
#> ✖ error target model
#> • end pipeline [1.733 seconds]
#> Error:
#> ! Error running targets::tar_make()
#>   Target errors: targets::tar_meta(fields = error, complete_only = TRUE)
#>   Tips: https://books.ropensci.org/targets/debugging.html
#>   Last error: missing values in object

writeLines(tar_traceback(model))
#> eval(expr = expr, envir = envir)
#> eval(expr = expr, envir = envir)
#> analyze_data(dataset1)
#> gls(model = outcome ~ factor, data = data, correlation = corSymm(form = ~measurement | unit), weights = v
#> tidy(., conf.int = TRUE, conf.level = 0.95)
#> gls(model = outcome ~ factor, data = data, correlation = corSymm(form = ~measurement | unit), weights = v
#> do.call(model.frame, mfArgs)
#> (function (formula, ...) 
#> UseMethod("model.frame"))(formula = ~measurement + unit + outcome + factor, dat
#> model.frame.default(formula = ~measurement + unit + outcome + factor, data = list(c(1, 1, 1, 1, 2, 2, 2, 
#> (function (object, ...) 
#> UseMethod("na.fail"))(list(c(1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 
#> na.fail.default(list(c(1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4, 1, 2, 3, 4
#> stop("missing values in object")
#> .handleSimpleError(function (condition) 
#> {
#>     state$error <- build_message(condition)
#>     state$tracebac
#> h(simpleError(msg, call))
```

**Reply by dipterix** [reply]
Is it possible just to just save the error object instead of the whole workspace? The workspace might be too big... We are analyzing GB-level data in-memory... I can resolve the error traceback and save it along with the error object.

**Reply by wlandau** [reply]
> Is it possible just to just save the error object instead of the whole workspace? The workspace might be too big... We are analyzing GB-level data in-memory... I can resolve the error traceback and save it along with the error object.

[Workspaces](https://books.ropensci.org/targets/debugging.html#workspaces) in `targets` are different from the `.RData` files from `save()`. These files are extremely lightweight, and they only save the text part of the traceback. (The raw output from `traceback()` can have sneaky environments with large data objects.)

**Reply by dipterix** [reply]
Thanks this is great :)
